### PR TITLE
Enable catkin build use_nmake on Windows

### PR DESCRIPTION
**Pull request by johnsonshih**
add missing part to handle use_nmake
update setup.bat template
skip replacing ':' with ';'

### CODE REVIEW

File: python/catkin/builder.py [hunk 1]
```diff
@@ -427,7 +427,7 @@ def build_catkin_package(
         make_executable = 'make'
 
     make_cmd = [make_executable]
-    make_cmd.extend(handle_make_arguments(make_args))
+    make_cmd.extend(handle_make_arguments(make_args, not use_nmake))
```
Comment (dirk-thomas) [inline python/catkin/builder.py]: Please pass keyword arguments by explicitly using the keyword: `handle_make_arguments(make_args, append_default_jobs_flags=not use_nmake)`.

Same below.

File: python/catkin/builder.py [hunk 2]
```diff
@@ -702,12 +704,13 @@ def build_package(
 
 
 def get_new_env(package, develspace, installspace, install, last_env, destdir=None):
+    env_script = 'env.bat' if platform.system() is 'Windows' else 'env.sh'
```
Comment (dirk-thomas) [inline python/catkin/builder.py]: Should the condition use `sys.platform == 'win32'` instead to not use `.bat` when using `cygwin`?

File: cmake/templates/setup.bat.in [hunk 3]
```diff
@@ -42,10 +42,21 @@ call %_SETUP_TMP%
...
+  set _CATKIN_ENVIRONMENT_HOOKS_%_HOOK_COUNT%_WORKSPACE=
+
+  REM call environment hook
+  call %%_CATKIN_ENVIRONMENT_HOOKS_%_HOOK_COUNT%%%
+  set _CATKIN_ENVIRONMENT_HOOKS_%_HOOK_COUNT%=
+
+  set CATKIN_ENV_HOOK_WORKSPACE=
+
+  set /a _HOOK_COUNT=%_HOOK_COUNT%+1
+  goto :hook_loop
```
Comment (dirk-thomas) [inline cmake/templates/setup.bat.in:59]: Can you please elaborate why this was changed.

File: python/catkin/builder.py (hunk 1 above)
Comment (johnsonshih) [reply python/catkin/builder.py]: use keyword argument

File: cmake/templates/setup.bat.in (hunk 3 above)
Comment (johnsonshih) [reply cmake/templates/setup.bat.in:59]: This is to match what the find_env_hooks() in _setup_util.py template (cmake/templates/_setup_util.py.in) does.  find_env_hooks() will put env hooks to environment variable _CATKIN_ENVIRONMENT_HOOKS_%d and _CATKIN_ENVIRONMENT_HOOKS_%d_WORKSPACE.  This batch file will look for those variables and invoke them

File: python/catkin/builder.py (hunk 2 above)
Comment (johnsonshih) [reply python/catkin/builder.py]: change to use sys.platform

File: cmake/all.cmake [hunk 4]
```diff
@@ -165,6 +169,9 @@ message(STATUS "catkin ${catkin_VERSION}")
 # ensure that no current package name is set
 unset(_CATKIN_CURRENT_PACKAGE)
 
+# tools/libraries.cmake
+configure_shared_library_build_settings()
```
Comment (dirk-thomas) [inline cmake/all.cmake:173]: Doesn't adding this change the current behavior? Before the option was not set which make `add_library` default to `STATIC`. With this line being added which basically runs this: https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/tools/libraries.cmake#L11-L13

the default behavior changes to `SHARED`.

Comment (johnsonshih) [reply cmake/all.cmake:173]: the BUILD_SHARED_LIBS is turned on in catkin/cmake/platform/windows.cmake, so it's ON by default (even on *nix).  In addition, configure_shared_library_build_settings() is called by catkin/cmake/catkin_workspace.cmake, too.  Which means BUILD_SHARED_LIBS is turned on by default in toplevel.cmake.  I'm not sure if add_library is default to STATIC, from my observation, it's default to SHARED.
  Do we have a consensus about the default value of this switch?  We need to nail down the default value and then change the code to make it consistent.

Comment (dirk-thomas) [reply cmake/all.cmake:173]: The `toplevel.cmake` / `catkin_workspace` is only being used by `catkin_make`. All other build tools (`catkin_make_isolated`, `catkin-tools`, `colcon`) don't use a workspace-level CMake file but invoke CMake for each package separately.

But I think you are right that `BUILD_SHARED_LIBS` is turned on by default. The reason is that both (one would already be sufficient) of the following files are included in the `all.cmake` file ([platform/windows.cmake](https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/all.cmake#L148), [tools/libraries.cmake](https://github.com/ros/catkin/blob/964f1bf385f356c8345677928801947d497a161e/cmake/all.cmake#L154)). Both declare the option and ensure that it is initialized with `on` if not already set.

Can you clarify why explicitly calling `configure_shared_library_build_settings()` here is necessary for making `nmake` work on Windows?

Comment (johnsonshih) [reply cmake/all.cmake:173]: For Windows, all shared libraries are built into Dll files and need to specify the storage-class attributes to indicate if a function/data is exported (the dll provides the function/data) or imported (the dll consumes the function/data).  We use macro to decide what attribute to use when compile a component (see https://github.com/ros/ros_comm/blob/melodic-devel/tools/rosbag_storage/include/rosbag/macros.h for example).  The macro checks the flag ROS_BUILD_SHARED_LIBS to decide the storage-class attribute.  configure_shared_library_build_settings() is where to decide whether to set the ROS_BUILD_SHARED_LIBS.

Comment (dirk-thomas) [reply cmake/all.cmake:173]: Thanks for clarifying - I forgot that this is what we use `ROS_BUILD_SHARED_LIBS` for.

**Comment by dirk-thomas** [comment]
Thank you for the improvement!

**Pull request merged**